go/login/login.go

```diff
@@ -0,0 +1,306 @@
+	V byte, hostIdType string, hostId string, action string) (*URLResponse, error) {
+	var prefix byte
+	var r URLResponse
+
+	d, e := userKey.Exchange(&serviceKey)
+	if e != nil {
+		return nil, e
+	}
+	r.d = d
+
+	r.Msg = Message{hostIdType, hostId, action}
+
+	if serviceKeyId == 0 {
+		// If no key ID was specified, send the first key byte as the ID.
+		prefix = serviceKey[0] & 0x7f
+	} else {
+		prefix = serviceKeyId & 0x7f
+	}
+	userPublic, e := userKey.Public()
+	if e != nil {
+		return nil, e
+	}
+	HandshakeInfo := Handshake{prefix, *userPublic, r.GetTag()}
+	r.HandshakeInfo = HandshakeInfo
+
+	return &r, nil
+}
+
+// Validates if the received tag corresponding to the base64-url encoded message constructed from the Message.
+func (r *URLResponse) ValidateAuthCode(tag string) bool {
+	return r.d.CheckEncoded(tag, r.Msg.Construct(), 0)
+}
+
+// Returns the tag corresponding to the Msg.
+func (r *URLResponse) GetTag() []byte {
+	return r.d.Tag(r.Msg.Construct(), 0)
+}
+
+// Returns a base64-encoded response token.
+func (r *URLResponse) GetEncToken() string {
+	return base64.URLEncoding.EncodeToString(r.GetTag())
+}
+
+// Client side login-lib handler.
+type Client struct {
+	ServiceKey   glome.PublicKey
+	UserKey      glome.PrivateKey
+	ServiceKeyId uint8
+	response     *URLResponse
+}
+
+// Client constructor. Sets Client.ServiceKey, Client.UserKey, Client.ServiceKeyId to the corresponding values
+// and Client.response to nil.
+func NewClient(sk glome.PublicKey, uk glome.PrivateKey, sId uint8) *Client {
+	return &Client{sk, uk, sId, nil}
+}
+
+func (c *Client) Construct(V byte, hostIdType string, hostId string, action string) (string, error) {
+	r, e := NewResponse(c.ServiceKeyId, c.ServiceKey, c.UserKey, V, hostIdType, hostId, action)
+	if e != nil {
+		return "", nil
+	}
+	c.response = r
+
+	var handshake = c.constructHandshake()
+	var msg = c.response.Msg.Construct()
+	var u = fmt.Sprintf("v%d/%s/", c.response.V, handshake)
+	if len(msg) > 0 {
+		u += fmt.Sprintf("%s/", msg)
+	}
+	return u, nil
+}
+
+func (c *Client) constructHandshake() string {
+	var handshake []byte
+	r := c.response
+
+	handshake = append(handshake, r.d.User[:]...)
+	handshake = append(handshake, r.GetTag()[:]...)
+	return base64.URLEncoding.EncodeToString(handshake[:])
+}
+
+// Validates if the received tag corresponding to the base64-url encoded message constructed from the Message.
+func (c *Client) ValidateAuthCode(tag string) bool {
+	return c.response.ValidateAuthCode(tag)
+}
+
+func (c *Client) getResponse() *URLResponse {
+	return c.response
+}
+
+type Server struct {
+	KeyFetcher func(uint8) glome.PrivateKey
+	UserKey    glome.PublicKey
+	response   *URLResponse
+}
+
+func NewServer(fetcher func(uint8) glome.PrivateKey, k glome.PublicKey) *Server {
+	return &Server{fetcher, k, nil}
+}
+
+func (s *Server) getResponse() *URLResponse {
+	return s.response
+}
+
+func (s *Server) ParseURLResponse(url *url.URL) (*URLResponse, error) {
+	path := url.Path
```

Comment:
You will probably need to use `RawPath` so that you can tell the difference between encoded slashes and real ones. https://golang.org/pkg/net/url/#URL

Reply:
Not actual anymore.